Implement new "welcome header"

philippotto opened the pull request:
The new welcome header with the empty-datasets-CTA currently looks like this:

![image](https://user-images.githubusercontent.com/2486553/106581915-b0e5da80-6543-11eb-9f4f-fa1df5f9f5a6.png)


In addition to the new header, I also removed the redundant headlines from the dashboard tab contents (as they merely repeated the tab titles).

Remaining todo:
- [x] add UI to hide the banner
- [x] persist whether the header was hidden by the user. the old banner was set via a GET parameter, which meant that the user would see this only once. should have a JSON-like property for a user to store stuff like this? @fm3 What do you think? for https://github.com/scalableminds/webknossos/issues/5022 we would need this, too. db migrations can update values within a json object, right?
- [x] clean up (e.g., remove code of old whats next banner)

### URL of deployed dev instance (used for testing):
- https://welcomeheader.webknossos.xyz

### Steps to test:
- log in on the dev instance
- check out the welcome banner in the dashboard and test its different actions
  - feel free to test the close-icon, but note that the header won't re-appear on page refresh. so, do this at the very end of testing :laughing: 
- checkout the empty dataset table with the overhauled CTA cards
  - they will only appear if the dataset list is empty. I hardcoded this for now to simplify testing.

### Issues:
- fixes #5023
- fixes #5020

------
- [ ] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Needs datastore update after deployment
- [x] Ready for review

fm3 commented:
Yes we can add a json field to the user, and yes we can edit those in SQL evolutions. I guess it should be stored with the multiuser, so that it is not reset when joining a new organization. What should it be called? `nuxInfo`?

philippotto commented:
> Yes we can add a json field to the user, and yes we can edit those in SQL evolutions.

Great!

> I guess it should be stored with the multiuser, so that it is not reset when joining a new organization.

Yes, I agree :+1: What is the current `/api/users/${newUser.id}` route referring to? Is it the multi user?

> What should it be called? `nuxInfo`?

Yes, sounds good to me :+1:

philippotto commented:
I decided to also incorporate the CTA in case of an empty datasets-table into this PR, since this should go well together with the general welcome-header.

The current version looks like this (for a completely empty account on wk.org):
![image](https://user-images.githubusercontent.com/2486553/106155533-12432d80-6181-11eb-8564-c3ce482cf47f.png)

The deployed version can be found here: https://welcomeheader.webknossos.xyz/?showWhatsNextBanner=true (notice that the `showWhatsNextBanner` will disappear after page load; so a reload will hide the banner. this will change once we can persist the information in the DB).

@hotzenklotz @normanrz Maybe you already want to give feedback on the wording?

fm3 commented:
@philippotto I just added the `novelUserExperienceInfos` field to the user json, defaulting to `{}`. When sent to the user update request, the field is completely overwritten for that user (or actually for the multiuser, as this is cross organization). Feel free to test it on this branch, does this work for you? :)

fm3 commented:
I’ll have a look 👀

hotzenklotz commented on frontend/javascripts/admin/welcome_ui.js:
```diff
@@ -0,0 +1,147 @@
+              />
+              <WhatsNextAction
+                title="Invite Your Colleagues"
+                description="Send invites to your colleagues that they can join your organization."
```
Find das etwas hackelig in der Formulierung. Evtl.
```suggestion
                description="Send invites to your colleagues and ask them to join your organization."
```

philippotto replied:
Whoops, I intended to write it as "send invites [...] SO that they can ...", but the word "so" was missing. Your suggestion sounds fine, too, though. I'll take yours :+1:

fm3 commented:
> However, the response contains novelUserExperienceInfos: {}. Do you know whats going on?

My bad, typical copy-paste error, should be fixed now :)

hotzenklotz commented on frontend/javascripts/admin/welcome_ui.js:
```diff
@@ -0,0 +1,147 @@
+              }}
+            >
+              Congratulations on your new webKnossos account! To hit the ground running, we
+              recommend the following steps for you:
```
Ich denke es heißt `recommend .. to you` https://dict.leo.org/german-english/recommend

```suggestion
              recommend the following steps to you:
```

daniel-wer commented on frontend/javascripts/admin/welcome_ui.js:
```diff
@@ -0,0 +1,147 @@
+              <WhatsNextAction
+                title="Open a Demo Dataset"
+                description="Have a look at a public dataset to experience webKnossos in action."
+                href={getDemoDatasetUrl()}
```
I think it could be confusing for non-demo users that they are redirected to webknossos.org, here. In the dataset list you implemented a switch to show the "Add Sample Dataset" Modal to non-demo users, instead. This could be a good solution here as well.

daniel-wer replied:
I just saw that you discussed this in Slack already and others agreed with the current solution. So feel free to ignore this :)

daniel-wer commented on conf/application.conf:
```diff
@@ -87,11 +87,16 @@ features {
+  # If isDemoInstance == true, `/createExplorative/hybrid/true` is appended to the URL so that a new tracing is opened.
+  # If isDemoInstance == false, `/view` is appended to the URL so that it's opened in view mode (since the user might not
+  # have an account).
+  publicDemoDatasetUrl = "https://webknossos.org/datasets/MPI_Brain_Research/2012-09-28_ex145_07x2_ROI2017_connectome"
```
The internal feedback was to use this dataset instead: https://webknossos.org/datasets/scalable_minds/l4dense_motta_et_al_demo

philippotto replied:
Good catch!

daniel-wer reviewed (commented):
Great work, massive design step up :1st_place_medal: 

Please see my small comments.

daniel-wer commented:
Also, thanks for making the testing on the deployed version so easy and frictionless! :grinning:

philippotto commented on frontend/javascripts/admin/admin_rest_api.js:
```diff
@@ -170,6 +170,26 @@ export function updateUser(newUser: $Shape<APIUser>): Promise<APIUser> {
+  };
+  const newUserAsync = updateUser(updateShape);
+
+  return [newUserSync, newUserAsync];
```
Yeah, I was also a bit surprised that we didn't need something similar earlier and was afraid I'd be missing something ^^ But I think the rest of our code is either quite atomic (via forms) or very "sophisticated" (via the save queue). So this landed somewhere in-between.

philippotto commented:
@fm3 I've got to bother with you two remaining things:

- the migration would need to be adapted so that existing users get `{"hasSeenDashboardWelcomeBanner": true}` as an initial value. I assume it's not as easy as changing it to `ALTER TABLE webknossos.multiUsers ADD COLUMN novelUserExperienceInfos JSONB NOT NULL DEFAULT '{"hasSeenDashboardWelcomeBanner": true}'::json;`, is it?
- hiding the banner didn't work for a non-admin during testing. I created a new account, activated that via the admin, and then clicked the close-icon. then, I get `You are not authorized to view or edit this resource`.  are users not allowed to update their own account?

fm3 commented:
> The migration would need to be adapted so that existing users get {"hasSeenDashboardWelcomeBanner": true} as an initial value. 

Done :) 

> Are users not allowed to update their own account?

Oh, yes that is actually correct 🤔 I’ll talk to youri today about how we can separate that.

youri-k commented on conf/evolutions/063-novelUserExperienceinfos.sql:
```diff
@@ -0,0 +1,32 @@
+DROP VIEW webknossos.multiUsers_;
+
+
+ALTER TABLE webknossos.multiUsers ADD COLUMN novelUserExperienceInfos JSONB NOT NULL DEFAULT '{}'::json;
```
Maybe cast the type directly to `jsonb`

youri-k commented on conf/evolutions/063-novelUserExperienceinfos.sql:
```diff
@@ -0,0 +1,32 @@
+ALTER TABLE webknossos.multiUsers ADD CONSTRAINT nuxInfoIsJsonObject CHECK(jsonb_typeof(novelUserExperienceInfos) = 'object');
+
+-- users that already existed *before* this change should not get the welcome banner
+UPDATE webknossos.multiUsers SET novelUserExperienceInfos = '{"hasSeenDashboardWelcomeBanner": true}'::json;
```
Same as above. Maybe cast the type directly to `jsonb`

youri-k reviewed (commented):
Backend LGTM apart from the small comment regarding the cast type in SQL

philippotto commented:
> Also, should non-admin members actually see the same welcome banner? They are not allowed to do most of what it explains?

I changed the code so that the invite-card is hidden when the user is not a team manager and the dataset-import-card is hidden if he is not a dataset manager. Admins can see all cards and "normal" users only the "open dataset" and "show tutorial video".

daniel-wer reviewed (commented):
Frontend LGTM. Should be ready to merge once the user editing permission issue is solved :)

fm3 commented:
I solved the permission issue by moving the novelUserExperienceInfos update to a separate route. @philippotto Could you check that this works for you? Also, I changed the api call in `admin_rest_api.js`, please have a look there too, I’m not sure if for example the `updateShape.id` field is still needed?

philippotto commented:
> I solved the permission issue by moving the novelUserExperienceInfos update to a separate route. @philippotto Could you check that this works for you? Also, I changed the api call in `admin_rest_api.js`, please have a look there too, I’m not sure if for example the `updateShape.id` field is still needed?

Awesome! The id field wasn't used anymore. I simplified the function a bit and it works great :man_dancing: 

Then, this should be good to go, right, @daniel-wer?

daniel-wer reviewed (approved):
> Awesome! The id field wasn't used anymore. I simplified the function a bit and it works great man_dancing
> 
> Then, this should be good to go, right, @daniel-wer?

I cannot see these changes yet, but once they're pushed, this should be ready to go :+1:

philippotto commented:
> > Awesome! The id field wasn't used anymore. I simplified the function a bit and it works great man_dancing
> > Then, this should be good to go, right, @daniel-wer?
> 
> I cannot see these changes yet, but once they're pushed, this should be ready to go

Hm, github seems to have some problems right now. I'll wait a bit.
